### DISCUSSION TITLE
Does dispatch happens after element morphing or before?

### DISCUSSION THREAD

**Discussion by whousehomestay** [livewire/livewire#7062] [1. Help]
What I want to achieve is performing some javascript codes after listening to an event dispatched from livewire, but I want to perform the codes just after the morphing / element updating process has been done on the livewire element.

For example, we want to add scrollInto behavior to element created using if-else in the livewire blade:

```
@if (session('status'))
    <div class="alert alert-success" id="success-alert">
        {{ session('status') }}
    </div>
@endif
```

on the livewire php file:

```
session()->flash('status', 'Post successfully updated.');
$this->dispatch('scroll-to-el', elementId: 'success-alert');
```

Then on javascript side, 
```
Livewire.on('scroll-to-el', ({elementId}) => {
document.getElementById(elementId).scrollIntoView({behavior: "smooth", block:"center"});
}
```

Using this code, sometimes it doesn't works because i think the event listener is executed before the element morphing process, so there is no element with id: 'success-alert' when the js code is executed.

So, does dispatch happens after element morphing or before? what the effective way to achieve this?

Sorry for my bad english.

**Comment by N1ebieski** [comment]
I'm afraid the author didn't foresee such a situation :/

I had a similar problem. My solution isn't pretty but works.

First, I created a state class:

```js
class Commit {
    constructor(responded) {
        this.responded = responded;
    }

    waitForRespond() {
        return new Promise((resolve) => {
            const checkRespond = () => {
                if (this.responded === true) {
                    resolve();
                } else {
                    setTimeout(checkRespond, 100);
                }
            };

            checkRespond();
        });
    }
}

const commit = new Commit(false);

export { commit as Commit };
```

Then I used [livewire commit hooks](https://livewire.laravel.com/docs/javascript#commit-hooks) to change the state:

```js
import { Commit } from "./commit";

Livewire.hook("commit.prepare", () => {
    Commit.responded = false;
});

Livewire.hook("commit", ({ respond }) => {
    respond(() => {
        Commit.responded = true;
    });
});
```

This way, if any of my livewire component dispatches an event, for example:

```php
$this->dispatch('my-event');
```

my another Alpine component listens:

```js
<x-my-alpine-component x-on:my-event.window="myListener" />

import { Commit } from "../livewire/commit";

export default function myAlpineComponent() {
    async myListener() {
        await Commit.waitForRespond();
    
        console.log('The livewire component has definitely been rendered');
    },
}
```

**Reply by whousehomestay** [reply]
Yeah maybe this is the solution for such problem. So it's basically listening for every livewire hook for 'commit.prepare' and 'commit' and marked a state variable for that, then if it has been commited, the action is commited? Yeah i think this is a good solution.
But, I'm just curious, I read on the doc if "commit" responded hook is run after a response is received but before it's processed, so, in order to mark the state variable when the livewire element manipulation is done, isn't we need to listen to "morph.updated" hook?
I'm confused because the doc isn't clear on this morphing lifecycle...

**Reply by N1ebieski** [reply]
The "respond" hook is enough for me, but you're probably right, it's better to use "succeed":

```js
    succeed(({ snapshot, effect }) => {
        // Runs after a successful response is received and processed
        // with a new snapshot and list of effects...
    })
```

**Comment by nikUnit** [comment]
` ```
```
if (!empty($date_from) || !empty($date_to) || !empty($search)) {
    <hook-for-update style="display: none" id="<?= uniqid() ?>"><?= $date_from . $date_to . $search ?></hook-for-update>
}
```
      <script>
          document.addEventListener('livewire:init', () => {
              Livewire.hook('morph.added', ({
                  el
              }) => {
                  console.log(el.tagName)
                  if (el.tagName === "HOOK-FOR-UPDATE") {
                      console.log(el.tagName)
                      your javascript scripts
                  }
              })
          })
      </script>
````
This code i added below my component, when properties changed i call hook where choose tagname then run my js scripts
